flink-core/src/main/java/org/apache/flink/configuration/Configuration.java

```diff
@@ -759,31 +791,49 @@ public boolean contains(ConfigOption<?> configOption) {
         }
 
         synchronized (this.confData) {
+            if (canBePrefixMap) {
+                removePrefixMap(this.confData, key);
+            }
             this.confData.put(key, value);
         }
     }
 
     private Optional<Object> getRawValue(String key) {
+        return getRawValue(key, false);
+    }
+
+    private Optional<Object> getRawValue(String key, boolean canBePrefixMap) {
         if (key == null) {
             throw new NullPointerException("Key must not be null.");
         }
 
         synchronized (this.confData) {
-            return Optional.ofNullable(this.confData.get(key));
+            final Object valueFromExactKey = this.confData.get(key);
+            if (!canBePrefixMap || valueFromExactKey != null) {
+                return Optional.ofNullable(valueFromExactKey);
+            }
+            final Map<String, String> valueFromPrefixMap =
+                    convertToPropertiesPrefixed(confData, key);
+            if (valueFromPrefixMap.isEmpty()) {
+                return Optional.empty();
+            }
+            return Optional.of(valueFromPrefixMap);
```

Comment:
```
            final Object valueFromExactKey = this.confData.get(key);
            if (valueFromExactKey == null && canBePrefixMap) {
                final Map<String, String> valueFromPrefixMap =
                        convertToPropertiesPrefixed(confData, key);
                if (valueFromPrefixMap.isEmpty()) {
                    return Optional.empty();
                }
                return Optional.of(valueFromPrefixMap);
            }
            return Optional.ofNullable(valueFromExactKey);
            ```

Reply:
(This is obviously highly opinionated, but in this specific case I prefer the current version better)

Reply:
Yes this one was not as obvious as the others as the negated form allowed for a quick escape. So I'd also be fine with the previous version.